processor/metricstransformprocessor/aggregate_label_values_operation.go

```diff
@@ -0,0 +1,79 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metricstransformprocessor
+
+import (
+	"fmt"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+)
+
+// aggregateOp aggregates the data points in metric based on given operation
+func (mtp *metricsTransformProcessor) aggregateLabelValuesOp(metric *metricspb.Metric, mtpOp mtpOperation) {
+	op := mtpOp.configOperation
+	labelSet := mtpOp.labelSetMap
+	// labelIdxs is a slice containing the indices of the selected labels.
+	// This is needed because label values are ordered in the same order as the labels
+	labelIdxs, _ := mtp.getLabelIdxs(metric, labelSet)
+	timeseriesGroup := mtp.groupTimeseriesWithNewValue(metric, labelIdxs, op.NewValue, mtpOp.aggregatedValuesSet)
+	// merge groups of timeseries
+	newTimeseries := mtp.aggregateTimeseriesGroups(timeseriesGroup, op.AggregationType, metric.MetricDescriptor.Type)
+	metric.Timeseries = newTimeseries
+}
+
+// groupTimeseries groups all timeseries in the metric that will be aggregated together based on the selected labels' values indicated by labelIdxs OR on the entire label values after replacing the aggregatedValues by newValue.
+// Depending on if newValue is set, the approach to group timeseries is different.
+// Returns a map from keys to groups of timeseries and a map from keys to the corresponding label values
+func (mtp *metricsTransformProcessor) groupTimeseriesWithNewValue(metric *metricspb.Metric, labelIdxs []int, newValue string, aggregatedValuesSet map[string]bool) map[string]*timeseriesGroupByLabelValues {
+	// key is a composite of the label values as a single string
+	keyToTimeseriesMap := make(map[string]*timeseriesGroupByLabelValues)
+	for _, timeseries := range metric.Timeseries {
+		var composedValues string
+		var newLabelValues []*metricspb.LabelValue
+		composedValues, newLabelValues = mtp.composeKeyWithNewValue(labelIdxs[0], timeseries, newValue, aggregatedValuesSet)
+		// group the timeseries together if their values match
+		timeseriesGroup, ok := keyToTimeseriesMap[composedValues]
+		if ok {
+			timeseriesGroup.timeseries = append(timeseriesGroup.timeseries, timeseries)
+		} else {
+			keyToTimeseriesMap[composedValues] = &timeseriesGroupByLabelValues{
+				timeseries:  []*metricspb.TimeSeries{timeseries},
+				labelValues: newLabelValues,
+			}
+		}
+	}
+	return keyToTimeseriesMap
+}
+
+// composeKeyWithNewValue composes the key for the timeseries with the aggregatedValues replaced by the newValue
+// Returns the key and a slice of the actual values used in this key
+func (mtp *metricsTransformProcessor) composeKeyWithNewValue(labelIdx int, timeseries *metricspb.TimeSeries, newValue string, aggregatedValuesSet map[string]bool) (string, []*metricspb.LabelValue) {
```

Comment:
What do you mean by "only returning keys & grouping info for labelsets where the value matched one of the expected values"? I don't think I quite understand. Right now what I am doing for this is that I look at all the label values for a timeseries, and substitute the values in the `aggregatedValuesSet` with the `NewValue`. Based on the value set after the substitution, I group the timeseries with the identital value set.

Reply:
Here's my thoughts:

Let's say the action is aggregate (sum) label values [a,b] for label x into new value A

And I have a metric with labels x,y,z.
- Label x has values a...z
- Label y has values a...z
- Label z has values a...z
- And for fun lets say the user has a bug, so there are two points with the same set of label values: x=z, y=z, z=z
==> 26^3 + 1 duplicate = 17,577 data points
(this metric might cause problems due to the high cardinality of label values)

What needs to be done:
- Group together all points where the value of x is a or b and the values of y,z are the same: There are 26^2 = 676 of these groups with 2 points each
- Merge these data points, ignore the rest

What your code will do:
- Group together all points where the value of x is a or b and the values of y,z are the same: There are 26^2 = 676 of these groups with 2 points each
- Group the two points with duplicate values x=z, y=z, z=z togther: This is 1 extra group with 2 points
- Group all other points into a single group of one point: There are 17,577 - 676*2 - 1*2 = 16,223 of these groups with 1 point each
- Merge these data points. In the case of a group with one point, you will create a new point, copy over the values (doing computations and iterations along the way)

The output will be the same except:
- The duplicated data point is mysteriously merged
- Some unnecessary work is done
- Personally I found it more complex to read as I was confused by why keys are created for things that don't need to be grouped. Conversely, maybe there is some elegance in creating a key for everything.

Note pretty much everything here also applies for the aggregate labels case as well, e.g. if you are aggregating away a label that has the same value for all data points, then nothing will actually get grouped / merged.

Reply:
So, how to proceed?

**Performance:**

One straightforward & obvious optimization is when you iterate through the `keyToTimeseriesMap`, if there is only one timeseries, then don't do anything - just put that timeseries the `newTimeSeriesList` directly (it's possible I missed that this is happening somewhere else already). While you could do better than this (e.g. update & remove from the time series list in place), I think that one small change would resolve most of my concerns around performance.

**Duplicated Data Points:**

- For the aggregate labels case, it feels okay to just merge these to me since the user is specifying how to aggregate data points generally.
- For the aggregate label values case, it's less clear that this is okay. The user has only stated that they want to aggregate points with particular values, and that aggregation strategy may not make sense for unrelated data points.

I don't have a good solution for now. Maybe we can ignore it?

**Readability:**

It's probably fine either way. I would need to think through the implications of ignoring points that we know are single values. You'd need to change quite a lot of the code to run in place which could add other complexity.

^ Please go away and have a think about all that and see what you think it the best strategy to take. Don't be afraid to defend your code if you think it makse sense how it is, but also don't be afraid to refactor if you think it will improve the code 😄

Reply:
Thanks so much for the detailed explaination! I do agree that for timeseries that I can directly determine to not be the target for this aggregation operation (e.g. the ones that don't even have label values in the aggregated value set), I can totally skip the process for creating a key for them and putting them in a group in the map, and later traverse the map to get them back out. Therefore, I have fixed the implementation to be, if it is a timeseries without any targeted label values, it is directly inserted into the new timeseries. This way, it would optimize the performance of the code, and also for the duplicate cases, they will be ignored by the aggregation process if they don't have targeted label values.